README.md
Клиент-серверное приложение для отправки уведомлений в Telegram.

**smsq** - клиент. Добавляет сообщения в очередь на отправку. Принимает следующие аргументы и опции:
*  --plaintext/-t - отключить интерпретацию разметки в сообщении
*  --low/-L - отправить сообщение с низким приоритетом (без звука)
*  --high/-H - отправить важное сообщение (звук включён, независимо от времени суток)
*  --debug/-d - включить вывод отладки в консоль
*  <queue> - имя очереди для добавления сообщения. Обязательный параметр
*  <message> - текст сообщения для отправки

Клиент ищет конфиг в следующих папках:
*  .
*  /etc
*  /spool

Минимальный конфиг должен содержать адрес сервера NATS:
```
{
    "nats_server": "nats://<nats.server.org>:4222"
}
```

Пример использования:
`smsq -t sms.hg_test.alerts.mon 'Hello, World!'`
Добавит сообщение "Hello, World!" в очередь "sms.hg_test.alerts.mon" как простой текст

**smss** - сервер. Слушает заданную очередь NATS и отправляет извлечённые сообщения в Telegram.
Конфиг:
```
{
    "tg_apitoken": "<Telegram API token>",
    "working_hours": [8, 22],
    "chats": {
        "sms.queue.project.subj": <chat/channel ID>
    },
    "nats_server": "nats://<nats.server.org>:4222",
    "nats_queue": "<queue>",
    "nats_project": "<project>"
}
```
Где:
*  tg_apitoken - Токен бота Telegram
*  working_hours - Рабочие часы, когда сообщения от бота будут приходить без звука
*  chats - список чатов/каналов/контактов для которых будут отправлены сообщения из очереди. Сообщения из очереди, которых нет в списке, будут проигнорированы
*  nats_server - адрес сервиса очередей NATS
*  nats_queue - очередь, часть имени адресата
*  nats_project - проект, часть имени адресата

Запуск:
`smss '*'`

Запустить сервер и слушать все сабжи из очереди (см. конфиг)
В данном случае, все события из очередей `sms.<queue>.<project>.*` будут отправлены в Telegram.

Можно слушать какой-то один subj:
`smss mon` -> `sms.<queue>.<project>.mon`

Для отладки сервер можно запустить с опцией **--debug**: `smss --debug mon`


**smsc** - отправляет сообщения напрямую в Telegram.

В конфиге sms.json должны быть явно перечисленны имена каналов/чатов/контактов
Очередь NATS не используется. Может сбоить из-за проблем с коннектом до серверов Telegram

**sms** - ранняя инкарнация проекта - "Всё в одном". Включает все вышеперечисленные части. Режим работы определяется набором опций:
*  --server - сервер очереди
*  --queue - клиент очереди

Или прямая отправка сообщения по умолчанию, если ни одной опции не указано


CHANGELOG
2019-10-11
 * Добавлены рабочие часы в конфиг
 * Включен вежливый режим по умолчанию
 * Добавлены опции -loud/-silent для явного задания режима оповещени

smsq/smsq.go
package main

import (
	"os"
	"log"
	"flag"
	"time"
	"strings"
	"encoding/json"

	"github.com/nats-io/nats.go"
)

type QMessage struct {
	PlainText    bool `json:"plaintext"`
	LowPriority  bool `json:"low_priority"`
	HighPriority bool `json:"high_priority"`
	MessageData  string `json:"data"`
}

type Config struct {
	NatsServer string `json:"nats_server"`
	// NatsQueue string `json:"nats_queue"`
	// NatsProject string `json:"nats_project"`
}

var config Config
var debug bool

func main() {
	var plaintext       bool
	var high_priority   bool
	var low_priority    bool

	// PARSE ARGUMENTS, PERFORM BASIC SANITY CHECKS ********** //
	flag.BoolVar(&debug,  "d", false, "Show some debug info")
	flag.BoolVar(&debug,  "debug",  false, "Show some debug info")

	flag.BoolVar(&plaintext,    "t", false, "Do not use any markup in the message")
	flag.BoolVar(&low_priority, "L", false, "Send silent/low priority message")
	flag.BoolVar(&high_priority,"H", false, "Send unmuted message nevertheless the gently policy")

	flag.BoolVar(&plaintext,    "plaintext",  false, "Do not use any markup in the message")
	flag.BoolVar(&low_priority, "low", false, "Send silent/low priority message")
	flag.BoolVar(&high_priority,"high",   false, "Send unmuted message nevertheless the gently policy")

	flag.Parse()

	args := flag.Args()

	if len(args) < 2 {
		log.Println("usage: smsq [--debug] <subject> <message>")
		os.Exit(15)
	}

	// FIND RIGHT CONFIG FILE TO READ ************************ //
	var filename string
	if _, err := os.Stat("sms.json"); err == nil {
		filename = "sms.json"

        } else if _, err := os.Stat("/spool/sms.json"); err == nil {
		filename = "/spool/sms.json"

        } else if _, err := os.Stat("/etc/sms.json"); err == nil {
		filename = "/etc/sms.json"

        }

	if debug { log.Println("CONFIG: " + filename) }

	// ******************************************************* //

	// READ CONFIG FROM KNOWN FILE *************************** //
	file, err := os.Open(filename)
	if err != nil {
		log.Print("CONFIG OPEN ERROR: ")
		log.Print(err)
		os.Exit(25)
	}

	decoder := json.NewDecoder(file)
	err = decoder.Decode(&config)
	if err != nil {
		log.Print("CONFIG PARSE ERROR: ")
		log.Print(err)
		os.Exit(27)
	}
	// ******************************************************* //

	msg_struct := QMessage {
		PlainText:    plaintext,
		LowPriority:  low_priority,
		HighPriority: high_priority,
		MessageData:  strings.Join(args[1:], " ")}

	if debug { log.Println("Encoding the message...") }
	msg_str, err := json.Marshal(msg_struct)
	if err != nil {
		log.Fatal(err)
		os.Exit(55)
	}

	if debug { log.Printf( "MODE: Queue\n" ) }

	opts := []nats.Option{nats.Name("NATS SMSQ Publisher")}
	opts = setupConnOptions(opts)

	// Connect to NATS
	if debug { log.Printf( "NATS [%s] Connecting...\n", config.NatsServer ) }
	nc, err := nats.Connect(config.NatsServer, opts...)
	if err != nil {
		log.Fatal(err)
	}

	defer nc.Close()

	msg_queue := args[0]

	if debug { log.Printf( "POSTing to [%s]...", msg_queue ) }
	nc.Publish(msg_queue, []byte(msg_str))
	nc.Flush()

	if err := nc.LastError(); err != nil {
		log.Fatal(err)
		os.Exit(75)
	}

	if debug { log.Printf( "POSTed\n" ) }

	// That's all folks!
	os.Exit(0)
}

func setupConnOptions(opts []nats.Option) []nats.Option {
	totalWait := 10 * time.Minute
	reconnectDelay := time.Second

	opts = append(opts, nats.ReconnectWait(reconnectDelay))
	opts = append(opts, nats.MaxReconnects(int(totalWait/reconnectDelay)))
	opts = append(opts, nats.DisconnectHandler(func(nc *nats.Conn) {
		log.Printf("NATS Disconnected: will attempt reconnects for %.0fm", totalWait.Minutes())
	}))
	opts = append(opts, nats.ReconnectHandler(func(nc *nats.Conn) {
		log.Printf("NATS Reconnected [%s]", nc.ConnectedUrl())
	}))
	opts = append(opts, nats.ClosedHandler(func(nc *nats.Conn) {
		log.Fatalf("NATS Exiting: %v", nc.LastError())
	}))
	return opts
}


smss/smss.go
package main

import (
	"os"
	"log"
	"flag"
	"time"
	"runtime"
	"encoding/json"

	"github.com/nats-io/nats.go"

	"github.com/go-telegram-bot-api/telegram-bot-api"

)

type QMessage struct {
	PlainText    bool `json:"plaintext"`
	LowPriority  bool `json:"low_priority"`
	HighPriority bool `json:"high_priority"`
	MessageData  string `json:"data"`
}

type Config struct {
	TgApitoken string `json:"tg_apitoken"`
	Chats map[string]int64 `json:"chats"`
	WorkingHours []int `json:"working_hours"`
	NatsServer string `json:"nats_server"`
	NatsQueue string `json:"nats_queue"`
	NatsProject string `json:"nats_project"`
}

var config Config
var debug bool

func main() {

	// PARSE ARGUMENTS, PERFORM BASIC SANITY CHECKS ********** //
	flag.BoolVar(&debug,  "d", false, "Show some debug info")
	flag.BoolVar(&debug,  "debug",  false, "Show some debug info")

	flag.Parse()

	args := flag.Args()

	if len(args) < 1 {
		log.Println("usage: smss [--debug] <subject|*>")
		os.Exit(15)
	}

	// FIND RIGHT CONFIG FILE TO READ ************************ //
	var filename string
	if _, err := os.Stat("sms.json"); err == nil {
		filename = "sms.json"

        } else if _, err := os.Stat("/spool/sms.json"); err == nil {
		filename = "/spool/sms.json"

        } else if _, err := os.Stat("/etc/sms.json"); err == nil {
		filename = "/etc/sms.json"

        }

	if debug { log.Println("CONFIG: " + filename) }

	// ******************************************************* //

	// READ CONFIG FROM KNOWN FILE *************************** //
	file, err := os.Open(filename)
	if err != nil {
		log.Print("CONFIG OPEN ERROR: ")
		log.Print(err)
		os.Exit(25)
	}

	decoder := json.NewDecoder(file)
	err = decoder.Decode(&config)
	if err != nil {
		log.Print("CONFIG PARSE ERROR: ")
		log.Print(err)
		os.Exit(27)
	}

	if len(config.WorkingHours) != 2 {
		// Working hours are't configured. Using default value: 8-22
		config.WorkingHours[0] = 8
		config.WorkingHours[1] = 22

		if debug { log.Println("No WorkingHours found. Failed to default ones: %d..%d\n", config.WorkingHours[0], config.WorkingHours[1]) }
	}

	var chat_id int64
	if debug { log.Printf( "CONNECTing Tg Server...\n" ) }

	bot, err := tgbotapi.NewBotAPI(config.TgApitoken)
	if err != nil {
		log.Fatal(err)
		os.Exit(15)
	}
	if debug { log.Printf( "CONNECTed!\n" ) }


	if debug { log.Printf( "MODE: Server\n" ) }

	// Connect Options.
	opts := []nats.Option{nats.Name("NATS SMSQ Publisher")}
	opts = setupConnOptions(opts)

	// Connect to NATS
	if debug { log.Printf( "NATS [%s] Connecting...\n", config.NatsServer ) }
	nc, err := nats.Connect(config.NatsServer, opts...)
	if err != nil {
		log.Fatal(err)
	}

	if debug { log.Printf( "NATS CONNECTed!\n" ) }

	msg_queue := getQueue(args[0], config)
	if debug { log.Printf( "QUEUE: %s\n", msg_queue ) }

	nc.Subscribe(msg_queue, func(msg *nats.Msg) {

		if val, ok := config.Chats[ msg.Subject ]; ok {
			chat_id = val

		} else {
			log.Printf("Uknown chat [%s]\n", msg.Subject)
			return
		}

		if debug { log.Printf( "MSG for %s / %d\n", msg.Subject, chat_id ) }

		// Try to decode a message
		var msg_struct QMessage
		err := json.Unmarshal(msg.Data, &msg_struct)
		if err != nil {
			log.Println(err)
			return;
		}

		tg_msg := tgbotapi.NewMessage(chat_id, msg_struct.MessageData)
		tg_msg.DisableNotification = silent_mode(msg_struct.LowPriority, msg_struct.HighPriority)

		if !msg_struct.PlainText {
			tg_msg.ParseMode = "Markdown"
		}

		if debug { log.Print( "POSTing... " ) }

		if _, err := bot.Send(tg_msg); err != nil {
			log.Println(err)
		}

		if debug { log.Print( "DONE!\n" ) }
	})
	nc.Flush()

	if err := nc.LastError(); err != nil {
		log.Fatal(err)
	}

	runtime.Goexit()

	if debug { log.Printf( "DISCONNECTED!\n" ) }
}

func silent_mode(Silent bool, Loud bool) bool {
	if Loud {
		if debug { log.Println("LOUD MODE OPTION") }
		return false

	} else if Silent {
		if debug { log.Println("SILENT MODE OPTION") }
		return true

	} else {
		dt := time.Now()
		hr, _, _ := dt.Clock()
		if hr < config.WorkingHours[0] || hr >= config.WorkingHours[1] {
			if debug { log.Println("SILENT MODE POLICY") }
			return true
		}
	}

	return false
}

func setupConnOptions(opts []nats.Option) []nats.Option {
	totalWait := 10 * time.Minute
	reconnectDelay := time.Second

	opts = append(opts, nats.ReconnectWait(reconnectDelay))
	opts = append(opts, nats.MaxReconnects(int(totalWait/reconnectDelay)))
	opts = append(opts, nats.DisconnectHandler(func(nc *nats.Conn) {
		log.Printf("NATS Disconnected: will attempt reconnects for %.0fm", totalWait.Minutes())
	}))
	opts = append(opts, nats.ReconnectHandler(func(nc *nats.Conn) {
		log.Printf("NATS Reconnected [%s]", nc.ConnectedUrl())
	}))
	opts = append(opts, nats.ClosedHandler(func(nc *nats.Conn) {
		log.Fatalf("NATS Exiting: %v", nc.LastError())
	}))
	return opts
}

func getQueue(target string, config Config) string {
	return "sms." + config.NatsProject + "." + config.NatsQueue + "." + target
}


smsc/smsc.go
package main

import (
	"os"
	"log"
	"flag"
	"time"
	"strings"
	"encoding/json"

	"github.com/go-telegram-bot-api/telegram-bot-api"

)

type Config struct {
	TgApitoken string `json:"tg_apitoken"`
	Chats map[string]int64 `json:"chats"`
	WorkingHours []int `json:"working_hours"`
	NatsServer string `json:"nats_server"`
	NatsQueue string `json:"nats_queue"`
	NatsProject string `json:"nats_project"`
}

var config Config
var debug bool

func main() {

	// PARSE ARGUMENTS, PERFORM BASIC SANITY CHECKS ********** //
	var loud bool
	var plain bool
	var silent bool

	flag.BoolVar(&loud,   "l", false, "Force to switch on sound even out of working hours")
	flag.BoolVar(&plain,  "p", false, "Force to do not use Markup in message")
	flag.BoolVar(&debug,  "d", false, "Show some debug info")
	flag.BoolVar(&silent, "s", false, "Send a silent message")

	flag.BoolVar(&loud,   "loud",   false, "Force to switch on sound even out of working hours")
	flag.BoolVar(&plain,  "plain",  false, "Force to do not use Markup in message")
	flag.BoolVar(&debug,  "debug",  false, "Show some debug info")
	flag.BoolVar(&silent, "silent", false, "Send a silent message")


	flag.Parse()

	args := flag.Args()

	if len(args) < 2 {
		log.Println("usage: ./sms [--debug|--loud|--silent] <chat_name> <message>")
		os.Exit(15)
	}

	// FIND RIGHT CONFIG FILE TO READ ************************ //
	var filename string
	if _, err := os.Stat("sms.json"); err == nil {
		filename = "sms.json"

        } else if _, err := os.Stat("/spool/sms.json"); err == nil {
		filename = "/spool/sms.json"

        } else if _, err := os.Stat("/etc/sms.json"); err == nil {
		filename = "/etc/sms.json"

        }

	if debug { log.Println("CONFIG: " + filename) }

	// READ CONFIG FROM KNOWN FILE *************************** //
	file, err := os.Open(filename)
	if err != nil {
		log.Print("CONFIG OPEN ERROR: ")
		log.Print(err)
		os.Exit(25)
	}

	decoder := json.NewDecoder(file)
	err = decoder.Decode(&config)
	if err != nil {
		log.Print("CONFIG PARSE ERROR: ")
		log.Print(err)
		os.Exit(27)
	}
	// ******************************************************* //

	msg_str := strings.Join(args[1:], " ")

	if debug { log.Printf( "CONNECTing Tg Server...\n" ) }

	bot, err := tgbotapi.NewBotAPI(config.TgApitoken)
	if err != nil {
		log.Fatal(err)
		os.Exit(15)
	}
	if debug { log.Printf( "CONNECTed!\n" ) }

	if debug { log.Printf( "MODE: Direct\n" ) }

	var chat_id int64
	if val, ok := config.Chats[args[0]]; ok {
		chat_id = val
	} else {
		log.Print("Uknown chat")
		os.Exit(25)
	}

	msg := tgbotapi.NewMessage(chat_id, msg_str)
	msg.DisableNotification = silent_mode(loud, silent, config)

	if !plain {
		msg.ParseMode = "Markdown"
	}

	if _, err := bot.Send(msg); err != nil {
		log.Println(err)
	}

}

func silent_mode(loud bool, silent bool, config Config) bool {
	if loud {
		if debug { log.Println("Loud mode used explicitly") }
		return true

	} else if !silent {
		if len(config.WorkingHours) != 2 {
			// Working hours are't configured. Using default value: 8-22
			config.WorkingHours[0] = 8
			config.WorkingHours[1] = 22

			if debug { log.Println("No WorkingHours found. Failed to default ones: %d..%d\n", config.WorkingHours[0], config.WorkingHours[1]) }
		}

		dt := time.Now()
		hr, _, _ := dt.Clock()
		if hr < config.WorkingHours[0] || hr >= config.WorkingHours[1] {
			if debug { log.Println("SILENT MODE POLICY") }
			return true
		}
	}

	return false
}
